cm/heatlearn/heatlearn.py

```diff
@@ -0,0 +1,328 @@
+import json
+import logging
+import os
+import subprocess  # nosec
+from time import sleep, time
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import rasterio
+import requests
+import urllib3
+from BaseCM.cm_output import validate
+from geocube.api.core import make_geocube
+from rasterio.mask import mask
+from shapely import wkt
+from shapely.geometry import Polygon, shape
+from shapely.ops import cascaded_union
+from tensorflow.keras.models import load_model
+
+logging.basicConfig(level=logging.INFO)
+
+# REST API for HDD
+if not os.path.exists("API_KEY.txt"):
+    raise FileNotFoundError("Missing API key.")
+with open("API_KEY.txt", "r") as f:
+    API_KEY = f.read().replace("\n", "")
+API_URL = "https://lab.idiap.ch/enermaps/api/rpc/enermaps_query_table"
+HEADERS = {"Authorization": "Bearer {}".format(API_KEY)}
+
+CURRENT_FILE_DIR = os.path.dirname(os.path.abspath(__file__))
+
+ESM_dict = {
+    1: 1,  # water
+    2: 2,  # railways
+    10: 3,  # NBU Area - Open Space
+    20: 4,  # NBU Area - Green ndvix
+    30: 5,  # BU Area - Open Space
+    40: 6,  # BU Area - Green ndvix
+    41: 7,  # BU Area - Green Urban Atlas
+    50: 0,  # BU Area - Built-up
+}
+
+MODELS = {
+    500: "9216fdce-9855-11eb-b3f9-3e22fb3fc3ab",
+    300: "2dcaa71c-9837-11eb-b3f9-3e22fb3fc3ab",
+}
+
+HDD_MODEL = {"slope": -0.0002855045732455094, "intercept": 1.8116549365250931}
+
+
+def replace_with_dict(ar: np.array, dic: dict = ESM_dict):
+    """Replace numpy values with dict (source: https://stackoverflow.com/a/47171600)."""
+    #
+    # Extract out keys and values
+    k = np.array(list(dic.keys()))
+    v = np.array(list(dic.values()))
+
+    # Get argsort indices
+    sidx = k.argsort()
+
+    # Drop the magic bomb with searchsorted to get the corresponding
+    # places for a in keys (using sorter since a is not necessarily sorted).
+    # Then trace it back to original order with indexing into sidx
+    # Finally index into values for desired output.
+    return v[sidx[np.searchsorted(k, ar, sorter=sidx)]]
```

Comment:
What are "ar, k, v, sidx"? Please give an understandable name to your variables.